File: tests/auto_test.rb
#!/usr/bin/env ruby

#
# Unit test K3 single-site execution.
# Automatically run a test on every SQL file in the DBToaster example directory
#
# Usage:
#    ./auto_test.rb [options]
#      -s (optional)        Test with only simple queries
#
#      -n NUMBER (optional) Test only a specific test number
#      -d Run distributed tests 
#      -l FILE run only a list of selected tests from a file

require 'optparse'
require 'pathname'

cur_path = File.expand_path(File.dirname(__FILE__))
examples_path = File.join(cur_path, "../external/dbtoaster/examples/queries/")

dir_filter = ""
test_num = nil
test_name = nil
distributed = false
test_list_name = nil
test_list = []
$num_nodes = 1
$q_type = "global"
$shuffle = false
$force_correctives = false
$order_file = ""

# option parser
opt_parser = OptionParser.new do |opts|
	opts.banner = "Usage: auto_test.rb [options]"
	opts.separator ""
	opts.separator "Specific options:"
	opts.on("-s", "--simple", "Do only simple tests") do
			dir_filter = "simple" 
		end
	opts.on("-l", "--filter [SIMPLE]", String, "Run only specific tests") do |s|
			dir_filter = s
		end
	opts.on("-f", "--file [NAME]", String, "Run a specific test file") do |s|
			test_name = s
		end
	opts.on("-n", "--testnum [NUMBER]", Integer, 
		  	"Choose a specific test") do |n|
		  test_num = n
		end
	opts.on("-d", "--dist", "Perform a distributed test") do
		  distributed = true
		end
    opts.on("-l", "--list [FILE]", String, 
          "Execute tests from a list in a file") do |file|
          test_list_name = file
        end
    opts.on("-p", "--nodes [NUMBER]", Integer,
            "Choose a number of nodes") do |n|
        $num_nodes = n
        end
    opts.on("-q", "--queue [STRING]", String,
            "Select type of queue: global/trigger/node") do |s|
        $q_type = s
        end
    opts.on("-x", "--shuffle", 
            "Shuffle the queues") do
        $shuffle = true
        end
    opts.on("-r", "--force", 
            "Force correctives") do
        $force_correctives = true
        end
    opts.on("-o", "--order [STRING]", String,
            "Use an order file instead of creating a trace") do |s|
        $order_file = s
        end
end

# now parse the options
opt_parser.parse!(ARGV)

# handle a test list
if test_list_name != nil then
    p = Pathname.new(test_list_name)
    (puts "Can't find list file #{test_list_name}"; exit) unless p.exist?
    File.open(test_list_name, "r").each_line do |line|
      s = /\d+/.match(line).to_s
      test_list << s.to_i
    end
end

p = Pathname.new(examples_path)
raise "Can't find path #{examples_path}" unless p.exist?

simple_files = []
all_files = []

# start off with simple dir so it appears first in our list
simple_path = File.join(examples_path, "simple")
simple_p = Pathname.new(simple_path)
raise "Can't find path #{simple_path}" unless simple_p.exist?

# add all simple files so they're first in our list
simple_p.children.each do |f|
  if f.extname == ".sql" then simple_files << ["simple", f.basename, f] end
end
simple_files.sort_by! do |_, name, _| name end

# Add all children but simple path files
p.children.each do |d|
  dir, last = d.split
  if last.to_s != "simple" then 
    d.children.each do |f|
        if f.extname == ".sql" then all_files << [last, f.basename, f] end
    end 
  end
end

# Sort the array
all_files.sort_by! do |last, fname, f| fname end
all_files.sort_by! do |last, fname, f| last end

# Combine with simple array
all_files = simple_files.concat(all_files)
all_files.map! do |_, _, f| f end

err_file = "./err_log.txt"
node_cmd = ""
$shuffle_cmd = if $shuffle then "--shuffle" else "" end
$force_cmd = if $force_correctives then "--force_correctives" else "" end

if distributed then
  test_cmd = File.join(cur_path, "./dist_test.rb")
  node_cmd = "-p #{$num_nodes} -q #{$q_type}"
  node_cmd = if $order_file=="" then node_cmd else "#{node_cmd} --order #{$order_file}" end
else
  test_cmd = File.join(cur_path, "./sql_test.rb")
end

# run either one test or many tests
if test_num == nil and test_name == nil then
	if File.exist?(err_file) then File.delete(err_file) end
	index = 1
	passed = 0
    tested = 0

	for file, last_dir in all_files
    if test_list == [] or (test_list.find_index(index) != nil) then
      long_name = file.to_s
      dir, file = File.split(long_name)
      first, last = File.split(dir)
      short_name = File.join(last, file)
      # Skip non-matching files
      if dir_filter != "" and dir_filter != last then 
        index += 1
        next 
      end
      print "Test #{index} (#{short_name}): "

      output = `#{test_cmd} #{node_cmd} #{$shuffle_cmd} #{$force_cmd} #{long_name}` 

      if (/ERROR|FAILED/ =~ output) != nil then 
        puts "ERROR"
        # record the error in one big file
        File.open(err_file, "a") do |out|
          out << "Test #{index} #{short_name}: ERROR\n#{output}\n\n"
        end
      else puts "PASSED"; passed += 1 
      end

      tested += 1
    end

      index += 1
	end
	puts "Passed #{passed} out of #{tested} tests"
elsif test_num != nil then # numbered test
	file = all_files[(test_num-1)]
	puts("Test #{test_num} (#{file.to_s}):")
    puts `#{test_cmd} #{node_cmd} #{$shuffle_cmd} #{$force_cmd} #{file.to_s}`
elsif test_name != nil then # named test
	puts("Test #{1} (#{test_name}):")
    puts `#{test_cmd} #{node_cmd} #{$shuffle_cmd} #{$force_cmd} #{test_name}`
end

File: tests/auto_test.py
#!/usr/bin/env python3

#
# Unit test k3-mosaic execution.
#
# Usage:
#    ./auto_test.py [options]


import argparse
import os
import sys

import dist_test

def check_exists(name, path):
    if not os.path.exists(path):
        raise Exception("No {0} found at {1}".format(name, path))

def run():
    parser = argparse.ArgumentParser()
    parser.add_argument('-p', '--path', action='store', dest='test_path',
                        default=None, help="Do all tests in a given directory")
    parser.add_argument('-f', '--file', action='append', dest='test_name',
                        help="Run a specific test file")
    parser.add_argument('-l', '--list', action='store', dest='test_list_name',
                        default=None, help="Execute tests from a list file")
    parser.add_argument('-d', '--dist', action='store_true', dest='distributed',
                        default=False, help="Distributed test")
    parser.add_argument('-n', '--nodes', action='store', type=int, dest='num_nodes',
                        default=1, help="Number of data nodes")
    parser.add_argument('-s', '--switches', action='store', type=int, dest='num_switches',
                        default=1, help="Number of switch nodes")
    parser.add_argument('-o', '--order', action='store', dest='order_file',
                        default=None, help="Use an order file instead of creating a trace")
    parser.add_argument('-v', '--verbose', action='store_true', dest='verbose',
                        default=False, help="See test results in detail")
    parser.add_argument('-m', '--map', action='store', dest='map_type',
                        default='set', help="Use vmap/multi maps")
    parser.add_argument('--no-new', action='store_false', dest='new_k3',
                        default=True, help="Create k3new file")
    parser.add_argument('--no-interp', action='store_false', dest='run_interp',
                        default=True, help="Run the interpreter")
    parser.add_argument('--workdir', action='store', type=str, dest='workdir',
                        default="temp", help="Directory to store work files")
    parser.add_argument('--gc-interval', action='store',
                        default=20000, type=int, help="Change gc interval (ms)")
    parser.add_argument('--msg-interval', action='store',
                        default=2, type=int, help="Change message interval (ms)")
    parser.add_argument('--no-log', action='store_false', dest='logging',
                        default=True, help="Disable logging")
    parser.add_argument('--no-multiidx', action='store_false', dest='multiidx',
                        default=True, help="Disable multi-index maps")
    parser.add_argument('--filemux', action='store_true', dest='filemux',
                        default=False, help="Enable filemux newprint")
    parser.add_argument('--safe-writes', action='store_true', dest='safe_writes',
                        default=False, help="Enable safe array writes")
    parser.add_argument('--overlap-factor', dest='map_overlap_factor',
                        default=None, help="Reduce layout of maps on nodes")
    parser.add_argument('--batch-size', dest='batch_size',
                        default=None, help="Size of batches")
    parser.add_argument('--debug', default=False, action='store_true', help="Debug output")
    parser.add_argument('--no-opt-route', default=True, action='store_false', dest='opt_route', help='disable optimized routing')
    parser.add_argument('--dump-info', default=False, action='store_true', help="Output for route generation")
    parser.add_argument('--print-warmup', dest='print_warmup', default=False, action='store_true', help="Print the warmup program")
    parser.add_argument('--trace', default=False, dest='do_trace', action='store_true', help="Tracing")
    parser.add_argument('--no-gen-deletes', action='store_false', dest='gen_deletes',
                        default=True, help="Don't generate delete triggers")
    parser.add_argument('--no-gen-correctives', action='store_false', dest='gen_correctives',
                        default=True, help="Don't generate correctives")
    parser.add_argument('--no-gen-single-vid', action='store_false', dest='gen_single_vid',
                        default=True, help="Don't generate non-isobatch mode")
    parser.add_argument('--run-correctives', action='store_true', dest='run_correctives',
                        default=False, help="Run with correctives")
    parser.add_argument('--no-run-isobatch', action='store_false', dest='run_isobatch',
                        default=True, help="Don't run in isobatch mode")
    args = parser.parse_args()

    test_list = []

    # handle a test list
    if args.test_list_name:
        check_exists('test list', args.test_list_name)
        # we need to add the file path to each file in the list
        list_path, _ = os.path.split(args.test_list_name)
        with open(args.test_list_name, 'r') as f:
            f_list = f.readlines()
            test_list += list(map(lambda x: os.path.join(list_path, x).replace('\n', ''),
                                  f_list))

    # take all files in a test_path
    if args.test_path:
        for root, dirs, files in os.walk(args.test_path):
            for f in files:
                if os.path.splitext(f)[1] == ".sql":
                    test_list += [os.path.join(root, f)]

    # handle test names from command line
    if args.test_name:
        test_list.extend(args.test_name)

    if len(test_list) == 0:
        parser.print_help()
        sys.exit(1)

    failed = 0
    verbose = args.verbose or len(test_list) == 1
    # run either one test or many tests
    for index, test_file in enumerate(test_list):
        print("[{0}] Testing {1}... ".format(index, test_file), end="")
        if verbose:
            print("")
        if not args.distributed:
            res = dist_test.run(test_file, verbose=verbose, distrib=False, logging=args.logging)
        else:
            res = dist_test.run(test_file,
                                num_nodes=args.num_nodes,
                                num_switches=args.num_switches,
                                order_file=args.order_file,
                                debug=args.debug,
                                verbose=verbose,
                                distrib=True,
                                new_k3=args.new_k3,
                                gen_correctives=args.gen_correctives,
                                gen_deletes=args.gen_deletes,
                                gen_single_vid=args.gen_single_vid,
                                run_correctives=args.run_correctives,
                                run_isobatch=args.run_isobatch,
                                map_type=args.map_type,
                                workdir=args.workdir,
                                run_interp=args.run_interp,
                                gc_interval=args.gc_interval,
                                msg_interval=args.msg_interval,
                                logging=args.logging,
                                filemux=args.filemux,
                                safe_writes=args.safe_writes,
                                map_overlap_factor=args.map_overlap_factor,
                                batch_size=args.batch_size,
                                opt_route=args.opt_route,
                                do_trace=args.do_trace,
                                dump_info=args.dump_info,
                                print_warmup=args.print_warmup
                                )
        # check if a test failed
        if not res:
            print("[ERROR]")
            failed += 1
        else:
            print("[PASSED]")

    if failed > 0:
        print("Failed {0}/{1} tests".format(failed, len(test_list)))
        sys.exit(1)
    else:
        print("Passed {0} test(s)".format(len(test_list)))

if __name__ == '__main__':
    run()
